File: java-kata/src/test/java/org/kata/sbe/loan/rules/LoanApplicationTestBuilder.java
package org.kata.sbe.loan.rules;

import org.kata.sbe.loan.domain.LoanApplication;

class LoanApplicationTestBuilder {
    private int creditScore = 750;
    private double debtToIncomeRatio = 0.35;
    private double loanToValueRatio = 0.75;
    private double annualIncome = 50000;
    private Double loanAmount = null;
    private Double propertyValue = null;


    public static LoanApplicationTestBuilder aValidApplication() {
        return new LoanApplicationTestBuilder();
    }

    public static LoanApplicationTestBuilder anApplicationWithLoanAmount(double loanAmount, double propertyValue) {
        return new LoanApplicationTestBuilder()
                .withLoanAmount(loanAmount)
                .withPropertyValue(propertyValue);
    }

    public LoanApplicationTestBuilder withCreditScore(int creditScore) {
        this.creditScore = creditScore;
        return this;
    }

    public LoanApplicationTestBuilder withDebtToIncomeRatio(double ratio) {
        this.debtToIncomeRatio = ratio;
        return this;
    }

    public LoanApplicationTestBuilder withLoanToValueRatio(double ratio) {
        this.loanToValueRatio = ratio;
        return this;
    }

    public LoanApplicationTestBuilder withAnnualIncome(double income) {
        this.annualIncome = income;
        return this;
    }

    public LoanApplicationTestBuilder withLoanAmount(double amount) {
        this.loanAmount = amount;
        return this;
    }

    public LoanApplicationTestBuilder withPropertyValue(double value) {
        this.propertyValue = value;
        return this;
    }

    public LoanApplication build() {
        LoanApplication.Builder builder = LoanApplication.builder()
                .withCreditScore(creditScore)
                .withDebtToIncomeRatio(debtToIncomeRatio)
                .withAnnualIncome(annualIncome);

        if (loanAmount != null && propertyValue != null) {
            builder.withLoanAmount(loanAmount)
                   .withPropertyValue(propertyValue);
        } else {
            builder.withLoanToValueRatio(loanToValueRatio);
        }

        return builder.build();
    }
}

File: dotnet-kata/Tests/LoanApplicationTestBuilder.cs
using Domain;

namespace Tests;

internal class LoanApplicationTestBuilder
{
    private int _creditScore = 750;
    private double _debtToIncomeRatio = 0.35;
    private double _loanToValueRatio = 0.75;
    private double _annualIncome = 50000;
    private double? _loanAmount;
    private double? _propertyValue;

    public static LoanApplicationTestBuilder AValidApplication()
        => new();

    public static LoanApplicationTestBuilder AnApplicationWithLoanAmount(double loanAmount, double propertyValue)
        => new LoanApplicationTestBuilder()
            .WithLoanAmount(loanAmount)
            .WithPropertyValue(propertyValue);

    public LoanApplicationTestBuilder WithCreditScore(int creditScore)
    {
        _creditScore = creditScore;
        return this;
    }

    public LoanApplicationTestBuilder WithDebtToIncomeRatio(double ratio)
    {
        _debtToIncomeRatio = ratio;
        return this;
    }

    public LoanApplicationTestBuilder WithLoanToValueRatio(double ratio)
    {
        _loanToValueRatio = ratio;
        return this;
    }

    public LoanApplicationTestBuilder WithAnnualIncome(double income)
    {
        _annualIncome = income;
        return this;
    }

    public LoanApplicationTestBuilder WithLoanAmount(double amount)
    {
        _loanAmount = amount;
        return this;
    }

    public LoanApplicationTestBuilder WithPropertyValue(double value)
    {
        _propertyValue = value;
        return this;
    }

    public LoanApplication Build()
    {
        var builder = LoanApplication.CreateBuilder()
            .WithCreditScore(_creditScore)
            .WithDebtToIncomeRatio(_debtToIncomeRatio)
            .WithAnnualIncome(_annualIncome);

        if (_loanAmount.HasValue && _propertyValue.HasValue)
        {
            builder.WithLoanAmount(_loanAmount.Value)
                   .WithPropertyValue(_propertyValue.Value);
        }
        else
        {
            builder.WithLoanToValueRatio(_loanToValueRatio);
        }

        return builder.Build();
    }
}